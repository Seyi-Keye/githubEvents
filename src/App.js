import React, { Component } from 'react';
import logo from './logo.svg';
import './App.css';
import request from 'axios';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider'
import {Card, CardHeader, CardTitle, CardText} from 'material-ui/Card';

class App extends Component {
  constructor(){
    super()
    this.state = {
      value: '',
      result: [],
      error: ''
    };
    this.handleChange = this.handleChange.bind(this);
    this.handleClick = this.handleClick.bind(this);
  }
  handleChange(event) {
    this.setState({value: event.target.value})
  }

  handleClick(){
    this.fetchUserEvent(this.state.value);
  }

  fetchUserEvent(value){
    request.get(`https://api.github.com/users/${value}/events`)
    .then((response) => {
      this.setState({result : response.data, error: ''});
    }, (error) => {
      this.setState({error: error.message});
    })
    .catch((error) => {console.log(error, 'anothe rerror')
      this.setState({error: error.message});
    })
  }

  displayTable() {
    const stuff = (this.state.error) ? (<p>{this.state.error}</p>) : (this.state.result.map((current) =>
        <div>
          <MuiThemeProvider>
            <Card>
                <CardHeader
                  title={current.created_at}
                  subtitle={current.repo.url}
                  avatar={current.actor.avatar_url}
                  key={current.id}
                />
                <CardTitle title={current.type} />
                <CardText>
                {current.type} by {current.actor.login}
                </CardText>
          </Card>
          </MuiThemeProvider>
        </div>
      ))
    return stuff;
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <img src={logo} className="App-logo" alt="logo" />
          <h1 className="App-title">Welcome to React</h1>
          <input placeholder='Enter github username' id='username' value={this.state.value} onChange={this.handleChange}/>
          <button onClick={this.handleClick}> Search</button>
        </header>
        {(!this.state.result.length) ? <p className="App-intro">Enter your Github Username above and the result will be displayed here</p> : this.displayTable() }
      </div>
    );
  }
}

export default App;
