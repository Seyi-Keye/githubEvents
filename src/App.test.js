import React from "react";
import ReactDOM from "react-dom";
import App from "./App";
import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";

jest.mock("axios", () => ({
  get: () =>
    Promise.resolve({
      data: [
        {
          id: "whatever_id",
          created_at: "eafeawf",
          repo: {
            url: "should be whatever"
          },
          actor: {
            avatar_url: "whatever"
          }
        }
      ]
    })
}));

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

const wrapper = shallow(<App />);

describe("<GithubEvent render method", () => {
  it("it renders a containing div", () => {
    expect(wrapper.find(".App")).toBeTruthy();
  });

  it("it renders a containing header div", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find(".App-header")).toBeTruthy();
  });

  it("it renders an image tag", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("img")).toBeTruthy();
  });

  it("it renders an h1 tag", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("h1")).toBeTruthy();
  });

  it("it renders an input tag", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("input")).toBeTruthy();
  });

  it("it renders a <button /> tag", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("button")).toBeTruthy();
  });

  it("it renders a <p /> tag", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.find("p")).toBeTruthy();
  });

  it("it renders a text", () => {
    const wrapper = shallow(<App />);
    expect(wrapper.text()).toContain(
      "Enter your Github Username above and the result will be displayed here"
    );
  });
});

describe("<Github Event functionalities", () => {
  it("contains 3 states", () => {
    const instanceState = {
      value: "",
      result: [],
      error: ""
    };
    expect(wrapper.state()).toEqual(instanceState);
  });

  it("test the content of the input", () => {
    const wrapper = shallow(<input value="I haver an input" />);

    expect(wrapper.find("input").props().value).toContain("I haver an input");
  });

  fit("simulate change event", () => {
    const handleChangeSpy = jest.spyOn(App.prototype, "handleChange");
    const wrapper = mount(<App />);
    const button = wrapper.find("input").first();
    button.simulate("change", {
      target: {
        value: "typing"
      }
    });
    expect(handleChangeSpy.mock.calls.length).toBe(1);
    expect(handleChangeSpy).toHaveBeenCalled();
    expect(wrapper.state().value).toBe("typing");
  });
  fit("simulate click event", cb => {
    const handleClickSpy = jest.spyOn(App.prototype, "handleClick");
    const wrapper = mount(<App />);
    console.log(wrapper.state(), "this are the states");
    const sampleData = {
      data: [
        {
          id: "whatever_id",
          created_at: "eafeawf",
          repo: {
            url: "should be whatever"
          },
          actor: {
            avatar_url: "whatever"
          }
        }
      ]
    };
    wrapper
      .find("button")
      .first()
      .simulate("click");
    console.log(
      wrapper.state(),
      "this are the states after, before setImmediate"
    );

    setImmediate(() => {
      expect(wrapper.state().result).toEqual(sampleData.data);
      cb();
    });

    // expect(handleClickSpy.calledOnce).toBeTruthy();
    expect(handleClickSpy).toHaveBeenCalled();
  });
});

describe("Snapshot test", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("displays card result", () => {
    const tree = renderer.create(<App />).toJSON();
    tree.children[0].children[2].props.value = "seyi";
    // manually trigger the callback
    tree.props.onChange;
    // re-rendering
    tree.children[0].children[3].props.onClick = component.toJSON();
    expect(tree).toMatchSnapshot();
    // manually trigger the callback
    tree.props.onClick;

    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("changes the state when input changes", () => {
    const component = renderer.create(<App />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the callback
    tree.props.onChange;
    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    // manually trigger the callback
    tree.props.onClick;

    // re-rendering
    tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
