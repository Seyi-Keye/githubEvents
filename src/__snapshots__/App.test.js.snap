// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test changes the state when input changes 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Welcome to React
    </h1>
    <input
      id="username"
      onChange={[Function]}
      placeholder="Enter github username"
      value=""
    />
    <button
      onClick={[Function]}
    >
       Search
    </button>
  </header>
  <p
    className="App-intro"
  >
    Enter your Github Username above and the result will be displayed here
  </p>
</div>
`;

exports[`Snapshot test changes the state when input changes 2`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Welcome to React
    </h1>
    <input
      id="username"
      onChange={[Function]}
      placeholder="Enter github username"
      value=""
    />
    <button
      onClick={[Function]}
    >
       Search
    </button>
  </header>
  <p
    className="App-intro"
  >
    Enter your Github Username above and the result will be displayed here
  </p>
</div>
`;

exports[`Snapshot test changes the state when input changes 3`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Welcome to React
    </h1>
    <input
      id="username"
      onChange={[Function]}
      placeholder="Enter github username"
      value=""
    />
    <button
      onClick={[Function]}
    >
       Search
    </button>
  </header>
  <p
    className="App-intro"
  >
    Enter your Github Username above and the result will be displayed here
  </p>
</div>
`;

exports[`Snapshot test renders correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h1
      className="App-title"
    >
      Welcome to React
    </h1>
    <input
      id="username"
      onChange={[Function]}
      placeholder="Enter github username"
      value=""
    />
    <button
      onClick={[Function]}
    >
       Search
    </button>
  </header>
  <p
    className="App-intro"
  >
    Enter your Github Username above and the result will be displayed here
  </p>
</div>
`;
